### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/Antonio27M-machine-learning#3]
Hola @Antonio27M , por aquí las issues!

## LO QUE HAS HECHO BIEN

- **Eliminación de duplicados correcta**: `df.drop_duplicates(inplace=True)` - Bien aplicado.

- **Visualización de la distribución de quality**: Buen uso de boxenplot.

- **Creación de la variable de calidad**: Usaste `np.select()` que es una forma válida de hacerlo.

- **Split con stratify**: ¡Muy bien! Usar `stratify=y` ayuda a mantener la proporción de clases.

- **Modelo KNN con datos escalados**: Entrenaste con datos escalados.

- **Hiperparametrización con ambos métodos**: RandomizedSearchCV y GridSearchCV.

---

### **1. Error CRÍTICO en el escalado - fit_transform en TEST**
```python
X_test_num = pd.DataFrame(norm_scaler.fit_transform(X_test[num_variables]), ...)  # ← ERROR
```

**NUNCA** uses `fit_transform` en los datos de test. Esto causa **data leakage** porque el scaler aprende de los datos de test.

**Corrección:**
```python
X_train_num = pd.DataFrame(norm_scaler.fit_transform(X_train[num_variables]), ...)
X_test_num = pd.DataFrame(norm_scaler.transform(X_test[num_variables]), ...)  # ← Solo transform
```

---

### **2. Mapeo de quality incorrecto**
```python
df["quality"] = np.select([df["quality"].isin([3, 4, 5]), df["quality"].isin([6, 7]), df["quality"].eq(8)], [0, 1, 2], ...)
```

El enunciado dice:
- 0 = Baja calidad (3, 4)
- 1 = Calidad media (5, 6)
- 2 = Alta calidad (7, 8)

Pero tú pusiste:
- 0 = (3, 4, **5**)  ← 5 debería ser media
- 1 = (6, **7**)  ← 7 debería ser alta
- 2 = (8)

---

### **3. NO hiciste la gráfica de accuracy vs k**
El enunciado pedía:
> *"Crea un bucle para probar diferentes valores de k. Grafica accuracy vs k"*

No hay ningún bucle ni gráfica.

---

### **4. NO usaste confusion_matrix ni classification_report**
El enunciado pedía evaluar con estas métricas además de accuracy.

---

### **5. NO guardaste el modelo**
No hay código con pickle para guardar el modelo.

---

### **6. Celda incompleta al final**

<img width="1022" height="430" alt="Image" src="https://github.com/user-attachments/assets/df42593a-5ff9-42fa-ba06-8a43d0f4ca89" />

Hay una celda con código incompleto.

---

**Antonio**, el ejercicio está bastante incompleto. El error más grave es el **data leakage** por usar `fit_transform` en test. Además te faltan la mayoría de los requisitos del enunciado: la gráfica de k, las métricas adicionales, la función de predicción (opcional) y guardar el modelo. También el mapeo de las categorías de calidad no sigue lo que indicaba el enunciado.
